### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [2]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data2 = pd.read_csv('deliveries.csv')
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

### Replacing `Deccan Chargers` with `Sunrisers Hyderabad`:

In [3]:
data1 = data1.replace("Deccan Chargers", "Sunrisers Hyderabad")

### Filtering SRH and CSK matches only:

In [12]:
kxip_srh = data1.loc[((data1['team1'] == 'Sunrisers Hyderabad') & (data1['team2'] == 'Kings XI Punjab') | (data1['team1'] == 'Kings XI Punjab') & (data1['team2'] == 'Sunrisers Hyderabad'))]
kxip_srh.head()

id  season        city        date                team1  \
18    19    2017   Hyderabad  2017-04-17  Sunrisers Hyderabad   
31    32    2017  Chandigarh  2017-04-28  Sunrisers Hyderabad   
76    77    2008   Hyderabad  2008-05-01  Sunrisers Hyderabad   
107  108    2008  Chandigarh  2008-05-23  Sunrisers Hyderabad   
150  151    2009   Kimberley  2009-05-09  Sunrisers Hyderabad   

               team2      toss_winner toss_decision  result  dl_applied  \
18   Kings XI Punjab  Kings XI Punjab         field  normal           0   
31   Kings XI Punjab  Kings XI Punjab         field  normal           0   
76   Kings XI Punjab  Kings XI Punjab         field  normal           0   
107  Kings XI Punjab  Kings XI Punjab         field  normal           0   
150  Kings XI Punjab  Kings XI Punjab         field  normal           0   

                  winner  win_by_runs  win_by_wickets   player_of_match  \
18   Sunrisers Hyderabad            5               0           B Kumar   
31   Sunrisers Hyderabad           26               0       Rashid Khan   
76       Kings XI Punjab            0               7          SE Marsh   
107      Kings XI Punjab            0               6          SE Marsh   
150      Kings XI Punjab            0               3  DPMD Jayawardene   

                                                 venue      umpire1  \
18           Rajiv Gandhi International Stadium, Uppal  AY Dandekar   
31   Punjab Cricket Association IS Bindra Stadium, ...  Nitin Menon   
76           Rajiv Gandhi International Stadium, Uppal  BR Doctrove   
107         Punjab Cricket Association Stadium, Mohali    Asad Rauf   
150                              De Beers Diamond Oval   GAV Baxter   

        umpire2 umpire3  
18   A Deshmukh     NaN  
31    CK Nandan     NaN  
76    RB Tiffin     NaN  
107    SJ Davis     NaN  
150   AM Saheba     NaN

### Question 1: Which team will win the match?

#### Solution: Checking counts of winner:

In [6]:
kxip_srh['winner'].value_counts()/kxip_srh.shape[0] * 100

Sunrisers Hyderabad    54.166667
Kings XI Punjab        45.833333
Name: winner, dtype: float64

**Answer:** As per the above counts, CSK won 54% matches against SRH. I would predict that CSK will win today's match.

## Question 2:

### Loading "Deliveries" dataset:

In [7]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

### Replacing `Deccan Chargers` with `Sunrisers Hyderabad`:

In [8]:
data2 = data2.replace("Deccan Chargers", "Sunrisers Hyderabad")

#### Solution: Filtering data with David Warner as batsman and calculating average per match strike rate:

In [9]:
dw = data2.loc[data2['batsman'] == 'DA Warner']
dw_balls = pd.DataFrame(dw.groupby('match_id')['ball'].count())
dw_runs = pd.DataFrame(dw.groupby('match_id')['batsman_runs'].sum())
dw_data = pd.merge(dw_balls, dw_runs, on='match_id')
dw_data['Strike Rate'] = 100 * dw_data['batsman_runs']/dw_data['ball']
dw_data = dw_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
dw_data

Balls  Runs  Strike Rate
match_id                          
1             9    14   155.555556
6            47    76   161.702128
10           34    49   144.117647
14           31    26    83.870968
19           57    70   122.807018
21            7     4    57.142857
25           42    43   102.380952
32           27    51   188.888889
36           60   126   210.000000
39           21    30   142.857143
43           34    40   117.647059
47            6     6   100.000000
52           52    69   132.692308
57           35    37   105.714286
141          40    51   127.500000
146          23    36   156.521739
150          32    21    65.625000
154          25    36   144.000000
158           7     4    57.142857
170           9    15   166.666667
172           1     0     0.000000
185           8     6    75.000000
189          33    57   172.727273
197          15    33   220.000000
200          72   107   148.611111
203           8     4    50.000000
209          22    33   150.000000
213           5     0     0.000000
218          11     6    54.545455
221          15    31   206.666667
224           2     0     0.000000
229           4     5   125.000000
238           3     1    33.333333
241          44    54   122.727273
250          28    46   164.285714
253          48    51   106.250000
259          50    77   154.000000
263          12     7    58.333333
266           7     2    28.571429
269           8     3    37.500000
274          15    13    86.666667
282           2     1    50.000000
289          25    21    84.000000
293          29    29   100.000000
301          10    19   190.000000
331          55   109   198.181818
357          19    21   110.526316
364           9     8    88.888889
369           9    14   155.555556
372           9    15   166.666667
374          46    79   171.739130
378           7     7   100.000000
380           7     3    42.857143
382          20    21   105.000000
385          58    77   132.758621
390          38    61   160.526316
394           4     0     0.000000
401          15    15   100.000000
404           4     1    25.000000
408           6     7   116.666667
412           1     0     0.000000
419          25    51   204.000000
424          42    66   157.142857
428          15     8    53.333333
433          18    13    72.222222
438           6     4    66.666667
443          36    40   111.111111
446          39    44   112.820513
451           6     2    33.333333
461          35    32    91.428571
466          12     8    66.666667
469          45    58   128.888889
474           2     0     0.000000
477          53    65   122.641509
481          51    61   119.607843
487           7     6    85.714286
489          11    12   109.090909
493          32    55   171.875000
496          24    44   183.333333
500          18    34   188.888889
503          47    59   125.531915
507          46    90   195.652174
511           3     4   133.333333
521          43    53   123.255814
525          28    57   203.571429
529          15    21   140.000000
531          21    28   133.333333
537          55    91   165.454545
541           7     9   128.571429
544          41    58   141.463415
551          28    61   217.857143
555           4     4   100.000000
558          19    24   126.315789
561          16    17   106.250000
564          54    81   150.000000
568          32    52   162.500000
572           5     6   120.000000
580          26    58   223.076923
584          14    13    92.857143
588          61    90   147.540984
591          50    74   148.000000
594          32    59   184.375000
598           4     0     0.000000
603          52    92   176.923077
610          18    24   133.333333
613          35    48   137.142857
616          15    11    73.333333
618          31    46   148.387097
622          42    52   123.809524
628          56    73   130.357143
631          16    18   112.500000
634          29    28    96.551724
635          59    9

#### Getting mean of the Strike rate  match wise:

In [10]:
avg_sr = dw_data['Strike Rate'].mean()
avg_sr

120.45423965546733

**Answer:** Average Strike Rate of David Warner in each match is 120.45.

## Question 3: What will be KL Rahul's contribution to team total?

### Filtering data where batsman is KLRahul and batting team is Kings XI Punjs:

In [15]:
df = data2.loc[(data2['match_id'] >= 7894) & (data2['batting_team'] == 'Kings XI Punjab')]
kxip_runs = pd.DataFrame(df.groupby('match_id')['total_runs'].sum())
kxip_runs = kxip_runs.rename({'total_runs' : 'kxip Runs'}, axis = 1)
rahul_runs = data2.loc[(data2['match_id'] >= 7894) & (data2['batsman'] == 'KL Rahul')]
rahul_runs = pd.DataFrame(rahul_runs.groupby('match_id')['batsman_runs'].sum())
rahul_runs = rahul_runs.rename({'batsman_runs' : "Rahul' Runs"}, axis = 1)
kxip_data = pd.merge(kxip_runs, rahul_runs, on = 'match_id')
kxip_data["Rahul' Contribution"] = kxip_data["Rahul' Runs"]/kxip_data['kxip Runs'] * 100
kxip_data

kxip Runs  Rahul' Runs  Rahul' Contribution
match_id                                             
7895            168           51            30.357143
7901            161           48            29.813665
7905            198           37            18.686869
7909            201           18             8.955224
7911            128           61            47.656250
7915            157           23            14.649682
7918            121           32            26.446281
7927            184           24            13.043478
7931            158           86            54.430380
7933            149          101            67.785235
7937            221           68            30.769231
7941             93           22            23.655914
7943            192           99            51.562500
7949            164            8             4.878049
11140           200            4             2.000000
11142           196            1             0.510204
11145           185           72            38.918919
11149           175           15             8.571429
11309           142           55            38.732394
11313           152           71            46.710526
11315           204          104            50.980392
11319           180           18            10.000000
11323           190           53            27.894737
11328           168           12             7.142857
11333           190           44            23.157895
11339           174           84            48.275862
11343           186            2             1.075269
11346           188           75            39.893617

### Average Rahul's contribution:

In [17]:
kxip_data["Rahul' Contribution"].mean()

27.37692857202169

**Answer:** As per the above analysis, we see that Rahul's contributon is around **27%** to total Punjab's score.

## Question 4: How many ball would SRH take to reach 50 runs on Team Total?

#### Solution: Observing match wise Rohit Sharma's runs:

In [19]:
srh_balls = data2.loc[data2['batting_team'] == 'Sunrisers Hyderabad']
srh_balls = pd.DataFrame(srh_balls.groupby('match_id')['total_runs'].sum())
srh_balls

total_runs
match_id            
1                207
6                140
10               158
14               155
19               159
21               191
25               176
32               207
36               209
39               185
43               136
47               140
52               158
57               128
64               110
66               142
68               214
72               155
77               164
80               165
85               148
89               140
92               181
98               182
102              153
108              175
111              153
114              147
121              104
124              184
128              168
131              169
136              148
140              141
144              100
147              145
151              168
155              166
158              161
163              166
164              133
171              158
172              154
174              143
175              150
179              190
186              170
189              171
195              148
199              131
204              157
207              115
210              157
214              186
216              139
220              151
225              178
229              145
232              104
233               82
236              137
240              154
245              175
249              165
253              168
260              135
265              129
272              146
275              149
279              175
286              136
292              135
295              138
300              198
313              119
316              138
323              157
328              196
331              187
337              126
341              177
346              100
348              186
352              150
356              181
359              145
366              190
373              128
376              132
384              126
388              150
394              115
397              132
402              119
405              127
414              159
416              144
423              130
428               81
432              161
435              146
440              150
444              178
449              136
453              132
455              132
461              133
466              121
469              184
474              145
477              172
481              155
487              134
489               44
493              157
496              205
500              142
503              161
507              189
511              160
521              164
525              172
529              127
531              163
537              176
541              137
544              150
551              192
555              132
558              201
561              163
564              185
568              135
572              113
580              182
584              142
588              145
591              137
594              146
598              118
603              194
610              129
613              177
616              137
618              146
622              180
628              158
631              149
634              162
635              163
636              208
7897             135
7900             155
7903             144
7909             191
7913             180
7916             130
7918             140
7921             152
7929             170
7932             155
7935             193
7939             192
7944             209
7947             183
7950             148
7952             179
7953             186
11138            189
11144            211
11147            233
11152            142
11310            102
11313            158
11321            121
11324            140
11329            167
11332            179
11336            169
11339            232
11342            172
11345            179
11413            171